PyRDF/Proxy.py

```diff
@@ -64,14 +106,62 @@ def GetValue(self):
             current action node in the computational graph.
 
         """
-        if not self.action_node.value:  # If event-loop not triggered
+        if not self.proxied_node.value:  # If event-loop not triggered
             from . import current_backend
-            generator = CallableGenerator(self.action_node.get_head())
+            generator = CallableGenerator(self.proxied_node.get_head())
             current_backend.execute(generator)
 
-        return self.action_node.value
+        return self.proxied_node.value
 
     def _call_handler(self, *args, **kwargs):
         # Handles an operation call to the current action node
         # and returns result of the current action node.
         return getattr(self.GetValue(), self._cur_attr)(*args, **kwargs)
+
+
+class TransformationProxy(Proxy):
+    """
+    A proxy object to an instantiated node. Used as a controller of the user
+    references to the node itself. When the user deletes reference to a
+    node (e.g. assigning the same variable to another operation), the proxy
+    object will get destroyed by Python, thus flagging the node to be without
+    user references anymore.
+    """
+
+    def __del__(self):
+        """Deletes current Proxy and flags the wrapped Node for pruning"""
+        self.proxied_node.has_user_references = False
+
+    def __getattr__(self, attr):
+        """
+        Intercepts calls on operation or attributes belonging to the proxied
+        node.
+
+        Returns either:
+        -------
+        function
+            If the attribute passed by the user is a supported operation, the
+            proxy will return a method to handle an operation call to the
+            current transformation node.
+
+        node attribute
+            If the attribute passed by the user is not an operation, the proxy
+            will try to return the corresponding attribute of the proxied node.
+        """
+
+        # Check if the parameter `attr` is an operation supported by
+        # the backend
+        from . import current_backend
+        if attr in current_backend.supported_operations:
+            # Stores the name of operation call in the node attributes
+            self.proxied_node._cur_attr = attr
+            return self.proxied_node._call_handler
+        else:
+            return getattr(self.proxied_node, attr)
+
+    @trackcalls
+    def __getstate__(self):
```

Comment:
should it raise a `pickle.PickleError()` or should we define a custom exception for that?

Reply:
Why do we need to care about pickling of `Proxy`s at all?

Reply:
Would you suggest to remove every reference to pickle in the `Proxy` class? 

With the current implementation proxies cannot be pickled without a proper implementation of `__setstate__` and `__getstate__` that overrides the Node's behaviour, indeed it leads to a known unstable state and an infinite recursion. This might change after moving the Node's methods to the Proxy class as you proposed above though.

Reply:
My point is that, if `Proxy`s are just an artifact and what contains the real information to unroll the execution is `Node`, we just need to serialize `Node`s.

Reply:
But we are only serializing `Node`s right now, because they contain all the information. If we instead provide the information on operations and arguments in `Proxy`s I don't know what that would imply as far as the serialization step is concerned

Reply:
I am not proposing to move any information that is needed for the execution to `Proxy`, that has to stay in `Node` of course.
But I think we are diverging from the topic of this particular discussion, which is why we need to serialize `Proxy`s. My point is that we don't and therefore no `__getstate__` is necessary.